AirCasting/SessionViews/ThresholdAlerts/ThresholdAlertsController.swift

```diff
@@ -0,0 +1,167 @@
+// Created by Lunar on 10/08/2022.
+//
+
+import Foundation
+import Resolver
+
+protocol ThresholdAlertsController {
+    func getAlerts(sessionUUID: SessionUUID, completion: @escaping (Result<[FetchedThresholdAlert], Error>) -> Void)
+    func processAlerts(delete toDelete: [DeleteAlertData], create toCreate: [NewThresholdAlertData], update toUpdate: [UpdateAlertData], completion: @escaping (Result<Void, ThresholdAlertsError>) -> Void)
+}
+
+enum ThresholdAlertsError: Error {
+    case failedRequests
+}
+
+struct DeleteAlertData {
+    let id: Int
+    let sensorName: String
+}
+
+struct UpdateAlertData {
+    let sensorName: String
+    let sessionUUID: SessionUUID
+    let oldId: Int
+    let newThresholdValue: Double
+    let newFrequency: ThresholdAlertFrequency
+}
+
+struct NewThresholdAlertData {
+    let sensorName: String
+    let sessionUUID: SessionUUID
+    let thresholdValue: Double
+    let frequency: ThresholdAlertFrequency
+}
+
+class DefaultThresholdAlertsController: ThresholdAlertsController {
+    let createAlertApiCommunitator: CreateThresholdAlertAPI = DefaultCreateThresholdAlertAPI()
+    let deleteAlertApiCommunitator: DeleteThresholdAlertAPI = DefaultDeleteThresholdAlertAPI()
+    let fetchAlertsApiCommunitator: FetchThresholdAlertsAPI = DefaultFetchThresholdAlertsAPI()
+    
+    func getAlerts(sessionUUID: SessionUUID, completion: @escaping (Result<[FetchedThresholdAlert], Error>) -> Void) {
+        fetchAlertsApiCommunitator.fetchAlerts { result in
+            switch result {
+            case .success(let existingAlerts):
+                let alertsOfTheSession = existingAlerts.filter({ $0.sessionUuid == sessionUUID.rawValue })
+                completion(.success(alertsOfTheSession))
+            case .failure(let error):
+                completion(.failure(error))
+            }
+        }
+    }
+    
+    func processAlerts(delete toDelete: [DeleteAlertData], create toCreate: [NewThresholdAlertData], update toUpdate: [UpdateAlertData], completion: @escaping (Result<Void, ThresholdAlertsError>) -> Void) {
+        var failed = false
```

Comment:
We used this approach in the app a couple of times already, but maybe it's time to switch to some better alternative. Could you tell me how this could be properly done?

Reply:
> 

@anna1901 It's a common approach, I'm also guilty of doing that. The reason is that it's almost impossible to hit a race condition with network callbacks. The intervals between server responses are **years** in the CPU frame of reference. But we always need to asses if that's the case - if those were CoreData calls for example I wouldn't be so sure about it.

To make it right here we'd probably add some lock (mutex) around this bool value